File: examples/ch12/ts/gatewayDecorators.ts
// Chapter 12 · Decorator: add behavior to a gateway without editing it.
//
// Each decorator implements IPaymentGateway and wraps ONE gateway, forwarding
// charge() while adding a cross-cutting concern. Decorators stack at runtime:
// new RetryingGateway(new LoggingGateway(new StripeGateway(), log)).

export interface Receipt {
  readonly provider: string;
  readonly amount: number;
  readonly confirmation: string;
}

export interface IPaymentGateway {
  charge(amount: number): Receipt;
}

export class TransientError extends Error {}

// The real gateway — the concrete component we wrap.
export class StripeGateway implements IPaymentGateway {
  charge(amount: number): Receipt {
    return { provider: "stripe", amount, confirmation: `ch_${Math.round(amount * 100)}` };
  }
}

// Base decorator: holds ONE wrapped gateway, forwards by default.
export class GatewayDecorator implements IPaymentGateway {
  constructor(protected readonly wrapped: IPaymentGateway) {}
  charge(amount: number): Receipt {
    return this.wrapped.charge(amount);
  }
}

export class LoggingGateway extends GatewayDecorator {
  constructor(
    wrapped: IPaymentGateway,
    private readonly log: string[],
  ) {
    super(wrapped);
  }
  override charge(amount: number): Receipt {
    this.log.push(`charging ${amount.toFixed(2)}`);
    const receipt = this.wrapped.charge(amount);
    this.log.push(`charged ${receipt.confirmation}`);
    return receipt;
  }
}

export class RetryingGateway extends GatewayDecorator {
  constructor(
    wrapped: IPaymentGateway,
    private readonly retries: number = 2,
  ) {
    super(wrapped);
  }
  override charge(amount: number): Receipt {
    let last: unknown;
    for (let attempt = 0; attempt <= this.retries; attempt++) {
      try {
        return this.wrapped.charge(amount);
      } catch (err) {
        if (!(err instanceof TransientError)) throw err; // only retry transients
        last = err;
      }
    }
    throw last; // retries exhausted
  }
}

File: examples/ch12/py/gateway_decorators.py
"""Chapter 12 · Decorator: add behavior to a gateway without editing it.

Each decorator implements IPaymentGateway and wraps ONE gateway, forwarding
charge() while adding a cross-cutting concern. Decorators stack at runtime:
RetryingGateway(LoggingGateway(StripeGateway())).
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass


@dataclass(frozen=True)
class Receipt:
    provider: str
    amount: float
    confirmation: str


class TransientError(Exception):
    """A failure worth retrying."""


class IPaymentGateway(ABC):
    @abstractmethod
    def charge(self, amount: float) -> Receipt: ...


class StripeGateway(IPaymentGateway):
    """The real gateway — the concrete component we wrap."""

    def charge(self, amount: float) -> Receipt:
        return Receipt("stripe", amount, f"ch_{int(round(amount * 100))}")


class GatewayDecorator(IPaymentGateway):
    """Base decorator: holds ONE wrapped gateway, forwards by default."""

    def __init__(self, wrapped: IPaymentGateway) -> None:
        self._wrapped = wrapped

    def charge(self, amount: float) -> Receipt:
        return self._wrapped.charge(amount)


class LoggingGateway(GatewayDecorator):
    """Augment: record around the delegated charge."""

    def __init__(self, wrapped: IPaymentGateway, log: list[str]) -> None:
        super().__init__(wrapped)
        self._log = log

    def charge(self, amount: float) -> Receipt:
        self._log.append(f"charging {amount:.2f}")
        receipt = self._wrapped.charge(amount)
        self._log.append(f"charged {receipt.confirmation}")
        return receipt


class RetryingGateway(GatewayDecorator):
    """Augment: retry the delegated charge on a transient failure."""

    def __init__(self, wrapped: IPaymentGateway, retries: int = 2) -> None:
        super().__init__(wrapped)
        self._retries = retries

    def charge(self, amount: float) -> Receipt:
        last: Exception | None = None
        for _ in range(self._retries + 1):
            try:
                return self._wrapped.charge(amount)
            except TransientError as exc:  # transient → try again
                last = exc
        assert last is not None
        raise last  # retries exhausted
